trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/witnesschecking/YamlWitnessProductAutomaton.java

```diff
@@ -0,0 +1,281 @@
+		}
+		return segments.size();
+	}
+
+	/** Checks if the waypoint matches with the statement. Assumption waypoints are not matched */
+	private boolean matchesWaypoint(final LETTER statement, final Waypoint waypoint) {
+		if (waypoint instanceof WaypointBranching && matchesStartLocation(statement, waypoint)) {
+			final ConditionAnnotation conditionAnnot = ConditionAnnotation.getAnnotation(statement);
+			return conditionAnnot != null
+					&& Boolean.parseBoolean(waypoint.getConstraint()) == !conditionAnnot.isNegated();
+		}
+		if (waypoint instanceof WaypointTarget) {
+			return matchesStartLocation(statement, waypoint);
+		}
+		if (waypoint instanceof WaypointFunctionEnter) {
+			return matchesEndLocation(statement, waypoint);
+		}
+		if (waypoint instanceof WaypointFunctionReturn) {
+			if (statement instanceof IIcfgReturnTransition) {
+				// function_return waypoints match the location of the function call
+				return matchesEndLocation(((IIcfgReturnTransition<?, ?>) statement).getCorrespondingCall(), waypoint);
+			}
+			return matchesEndLocation(statement, waypoint);
+		}
+		return false;
+	}
+
+	/** Returns true if the start line and column of the statement match the location of the waypoint */
+	private static boolean matchesStartLocation(final IElement statement, final Waypoint waypoint) {
+		final ILocation programLoc = ILocation.getAnnotation(statement);
+		final Location witnessLoc = waypoint.getLocation();
+		return programLoc.getStartLine() == witnessLoc.getLine()
+				&& (witnessLoc.getColumn() == null || witnessLoc.getColumn() == programLoc.getStartColumn());
+	}
+
+	/** Returns true if the end line and column of the statement match the location of the waypoint */
+	private static boolean matchesEndLocation(final IElement statement, final Waypoint waypoint) {
+		final ILocation programLoc = ILocation.getAnnotation(statement);
+		final Location witnessLoc = waypoint.getLocation();
+		return programLoc.getEndLine() == witnessLoc.getLine()
+				&& (witnessLoc.getColumn() == null || witnessLoc.getColumn() == programLoc.getEndColumn());
+	}
+
+	/** Creates a ProductPredicate with the counters */
+	private IPredicate createProductState(final IPredicate predicate, final int waypointCounter,
+			final int violationSequenceCounter) {
+		final ProductPredicate result = new ProductPredicate(predicate, waypointCounter, violationSequenceCounter);
+		mAllProductStates.add(result);
+		return result;
+	}
+
+	/**
+	 * IPredicate annotated with counters for the Violation Sequence and Segments of an entry-based violation witness
+	 */
+	private static final class ProductPredicate extends AnnotatedPredicate<IPredicate, Pair<Integer, Integer>>
+			implements ISLPredicate {
+		public ProductPredicate(final IPredicate underlying, final int segmentCounter,
+				final int violationSequenceCounter) {
+			super(underlying, new Pair<>(segmentCounter, violationSequenceCounter));
```

Comment:
Would it be more convenient to store the violation sequence directly rather than this `violationSequenceCounter`?

Reply:
Maybe slightly, but the full violation sequences are not necessary outside of this class and they might need much more memory than just storing some counters.

Reply:
But you're just storing the pointer to the object (which must still exist, otherwise successor computation fails), so it's not saving any significant amount of memory. But it's not really important either way :shrug: